sdr_hackrf.c

```diff
@@ -0,0 +1,289 @@
+
+    if (len != Modes.sdr_buf_size) {
+        fprintf(stderr, "weirdness: hackRF gave us a block with an unusual size (got %u bytes, expected %u bytes)\n",
+                (unsigned) len, (unsigned) Modes.sdr_buf_size);
+        if (len > Modes.sdr_buf_size) {
+            unsigned discard = (len - Modes.sdr_buf_size + 1) / 2;
+            outbuf->dropped += discard;
+            buf += discard * 2;
+            len -= discard * 2;
+        }
+    }
+    
+    if (was_odd) {
+        ++buf;
+	--len;
+	++outbuf->dropped;
+    }
+
+    was_odd = (len & 1);
+    slen = len / 2; // Drops any trailing odd sample, that's OK
+
+    if (free_bufs == 0 || (dropping && free_bufs < MODES_MAG_BUFFERS / 2)) {
+        // FIFO is full. Drop this block.
+        dropping = 1;
+        outbuf->dropped += slen;
+        sampleCounter += slen;
+        // make extra sure that the decode thread isn't sleeping
+        unlockReader();
+        return 1;
+    }
+
+    dropping = 0;
+    unlockReader();
+
+    // Compute the sample timestamp and system timestamp for the start of the block
+    outbuf->sampleTimestamp = sampleCounter * 12e6 / Modes.sample_rate;
+    sampleCounter += slen;
+    // Get the approx system time for the start of this block
+    block_duration = 1e3 * slen / Modes.sample_rate;
+
+    outbuf->sysTimestamp = mstime();
+    outbuf->sysMicroseconds = mono_micro_seconds();
+
+    outbuf->sysTimestamp -= block_duration;
+    outbuf->sysMicroseconds -= block_duration * 1000;
+
+    // Copy trailing data from last block (or reset if not valid)
+    if (outbuf->dropped == 0) {
+        memcpy(outbuf->data, lastbuf->data + lastbuf->length, Modes.trailing_samples * sizeof (uint16_t));
+    } else {
+        memset(outbuf->data, 0, Modes.trailing_samples * sizeof (uint16_t));
+    }
+
+    // Convert the new data
+    outbuf->length = slen;
+    hackRF.converter(buf, &outbuf->data[Modes.trailing_samples], slen, hackRF.converter_state, &outbuf->mean_level, &outbuf->mean_power);
+    // Push the new data to the demodulation thread
+    lockReader();
+
+    Modes.mag_buffers[next_free_buffer].dropped = 0;
+    Modes.mag_buffers[next_free_buffer].length = 0; // just in case
+    Modes.first_free_buffer = next_free_buffer;
+
+    // accumulate CPU while holding the mutex, and restart measurement
+    end_cpu_timing(&thread_cpu, &Modes.reader_cpu_accumulator);
+    start_cpu_timing(&thread_cpu);
+
+    wakeDecode();
+    unlockReader();
+
+    return 0;
+}
+
+void hackRFRun() {
+    if (!hackRF.device) {
+        return;
+    }
+
+    start_cpu_timing(&thread_cpu);
+
+    int status;
+    if ((status = hackrf_start_rx(hackRF.device, hackrfCallback, NULL)) != HACKRF_SUCCESS) {
+        fprintf(stderr, "hackrf_start_rx failed: %s\n", hackrf_error_name(status));
+    }
+
+    struct timespec ts;
+    clock_gettime(CLOCK_REALTIME, &ts);
+    while (!Modes.exit) {
+        threadTimedWait(&Threads.reader, &ts, 50);
```

Comment:
While loop is needed since hackrf_start_rx returns immediately. Not sure of the best way to handle waiting, since the HackRF API doesn't provide a blocking function similar to bladerf_stream or rtlsdr_read_sync.

Reply:
https://github.com/greatscottgadgets/hackrf/blob/master/host/libhackrf/src/hackrf.c#L1976